src/geometry/types.rs

```diff
@@ -122,57 +127,63 @@ impl Vector {
     }
 }
 
-// Rectangle /////////////////////////////////////////////////////////////////
-#[derive(PartialEq, Debug)]
+// Size //////////////////////////////////////////////////////////////////////
+
+#[derive(Clone, PartialEq, Debug)]
 pub struct Size {
     pub width: f32,
     pub height: f32,
 }
 
-#[derive(PartialEq, Debug)]
+impl Size {
+    pub fn new(width: f32, height: f32) -> Size {
+        if width <= 0.0 || height <= 0.0 {
+            panic!("A size's width and/or height must be positive.");
+        }
+        Size {
+            width,
+            height,
+        }
+    }
+}
+
+// Rect //////////////////////////////////////////////////////////////////////
+//
+// A rectangle whose origin denotes the position of the bottom left corner.
+
+#[derive(Clone, PartialEq, Debug)]
 pub struct Rect {
-    // the leftmost coordination system of the rectangle
     pub origin: Point,
-    // the width and high of the rectangle
     pub size: Size,
 }
 
 impl Rect {
-    fn quarter_sized(&self) -> Rect {
-        Rect{
-            origin: Point{
-                x: self.origin.x,
-                y: self.origin.y,
-            },
-            size: Size{
-                width: self.size.width/ 2.0,
-                height: self.size.height/2.0,
-            }
+    pub fn new(x: f32, y: f32, width: f32, height: f32) -> Rect {
+        Rect {
+            origin: Point::new(x, y),
+            size: Size::new(width, height),
         }
     }
 
-    pub fn quadrants(&self) -> (Rect, Rect, Rect, Rect) {
-
-        //the upper left rectangle
-        let mut upper_left = self.quarter_sized();
-
-        //the upper right rectangle
-        let mut upper_right = self.quarter_sized();
+    fn quarter_sized(&self) -> Rect {
+        let (w, h) = (self.size.width / 2.0, self.size.height / 2.0);
+        Rect::new(self.origin.x, self.origin.y, w, h)
+    }
 
-        upper_right.origin.x = upper_right.origin.x + upper_right.size.width;
+    pub fn quadrants(&self) -> (Rect, Rect, Rect, Rect) {
+        let lower_left = self.quarter_sized();
```

Comment:
is the coordinator of lower_left here not (x, y+ height/2) ?

Reply:
I thought about it, it actually makes more sense to put the origin of the `Rect` on the lower left, since the coordinate system that we're already using puts the origin in the center. In other words, a point at (0, 0) will be the _center_ of our environment. So a `Rect` positioned at this point that has a width and height of 3 will in the positive quadrant.